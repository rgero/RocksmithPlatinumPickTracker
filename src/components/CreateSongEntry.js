import React from 'react';
import {connect} from 'react-redux';
import SongForm from './SongForm';
import { addSong } from '../actions/songs';

export class CreateSongEntry extends React.Component {
    constructor(props){
        super(props);
        this.onSubmit = this.onSubmit.bind(this);
    }

    onSubmit(song){
        this.props.onSubmit(song);
        this.props.history.push('/')
    }

    render(){
        return(
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Song</h1>
                    </div>
                </div>
                <div className="content-container">
                    <SongForm onSubmit={this.onSubmit} />
                </div>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        onSubmit: (song) => dispatch(addSong(song))
    }
}

export default connect(undefined, mapDispatchToProps)(CreateSongEntry); // Check out the react-redux documentation to understand the connect statement here.