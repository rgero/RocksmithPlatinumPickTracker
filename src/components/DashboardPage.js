import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import PickListFilter from './PickListFilters';
import PickList from './PickList';




const DashboardPage = () => (
    (
        <div>
            <div className='content-container'>
                <div className='content-navbar'>
                    <Link className='button' to='/import'>Import Data</Link>
                    <Link className='button' to='/create'>Log New Pick</Link>
                </div>
                <PickListFilter />
                <PickList />
            </div>
        </div>
    )
);

export default DashboardPage;