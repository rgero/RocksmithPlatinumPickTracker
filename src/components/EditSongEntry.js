import React from 'react';
import {connect} from 'react-redux';
import SongForm from './SongForm';
import {startEditSong, startRemoveSong} from '../actions/songs';

export class EditSongPage extends React.Component {
    constructor(props){
        super(props);
        this.onSubmit = this.onSubmit.bind(this);
        this.removeItem = this.removeItem.bind(this);
    }

    onSubmit(song){
        this.props.onSubmit(this.props.song.id, song)
        this.props.history.push('/')
    }

    removeItem(){
        this.props.removeItem(this.props.song.id);
        this.props.history.push('/')
    }

    render(){
        return(
            <div>
                <div className='page-header'>
                    <div className='content-container'>
                        <h1 className='page-header__title'>Editing '{this.props.song.artist} - {this.props.song.songName}'</h1>
                    </div>
                </div>
                <div className='content-container'>
                    <SongForm
                        songInfo = {this.props.song}
                        onSubmit={this.onSubmit}
                        
                    />
                    <button className='button button--secondary' onClick={this.removeItem}>Remove</button>
                </div>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        onSubmit: (id, song) => { 
            dispatch(startEditSong(id, song))
        },
        removeItem: (id) => dispatch(startRemoveSong(id))
    }
}

const mapStateToProps = (state, props) => {
    return {
        song: state.songs.find((song) => song.id === props.match.params.id)
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(EditSongPage);