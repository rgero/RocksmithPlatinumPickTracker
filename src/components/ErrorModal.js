import React from 'react';
import Modal from 'react-modal';

const ErrorModal = (props) =>
    (
        <Modal
            isOpen={!!props.errorMessage}
            contentLabel='Error with Form' // This is for people who have accessiblity options enabled.
            onRequestClose={props.clearError} // This is fired if you click in the background or the escape key.
            closeTimeoutMS={200}
            className='modal'
        >
            {props.errorMessage && 
                <div>
                    <h2>Error</h2>
                    <div>{props.errorMessage}</div>
                    <div><button className='button' onClick={props.clearError}>Close</button></div>
                </div>
            }
        </Modal>
    )

export default ErrorModal