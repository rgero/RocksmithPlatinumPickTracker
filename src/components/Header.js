import React from 'react';
import { Link } from 'react-router-dom';
import {connect} from 'react-redux';
import {startLogout} from '../actions/auth';

export const Header = ({ startLogout }) => (
    <header className="header">
        <div className="header__content">
            <Link className="header__title" to="/practices" exact="true"><h1>Rocksmith Pick Tracker</h1></Link>
            <button className="button button--link" onClick={startLogout}>Log out</button>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({

    startLogout: () => dispatch(startLogout())
})

export default connect(undefined, mapDispatchToProps)(Header);