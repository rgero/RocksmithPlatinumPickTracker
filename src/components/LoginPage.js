import React from 'react';
import {connect} from 'react-redux';
import {startGoogleLogin} from '../actions/auth';

export class LoginPage extends React.Component {

    constructor(props){
        super(props);
    }

    render() {
        return (
            <div className="box-layout">
                <div className="box-layout__box">
                    <h1 className="box-layout__title">Welcome to the Rocksmith Pick Tracker</h1>
                    <p>It's time to write a tag-line</p>
                    <button className="button" onClick={this.props.startGoogleLogin}>Login with Google!</button>
                </div>
            </div>
        )
    }
};

const mapDispatchToProps = (dispatch) => ({

    startGoogleLogin: () => dispatch(startGoogleLogin())
})

export default connect(undefined, mapDispatchToProps)(LoginPage);