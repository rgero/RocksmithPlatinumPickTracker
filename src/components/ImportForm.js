import React from 'react';
import ErrorModal from './ErrorModal';

export default class ImportForm extends React.Component {

    constructor(props)
    {
        super(props);

        this.onSubmit = this.onSubmit.bind(this);
        this.clearError = this.clearError.bind(this);
        this.onFileTypeChange = this.onFileTypeChange.bind(this);
        this.onChangeFile = this.onChangeFile.bind(this);
        this.handleCheckChange = this.handleCheckChange.bind(this);

        this.state = {
            error: "",
            filePath: "",
            fileType: "csv",
            hasHeaders: false
        }
    }

    onChangeFile(event) {
        event.stopPropagation();
        event.preventDefault();
        var file = event.target.files[0];
        this.setState({filePath: file}); /// if you want to upload latter
    }

    onFileTypeChange(e){
        const value = e.target.value
        this.setState(()=>({
            fileType: value
        }))
    }

    handleCheckChange(e){
        const value = e.target.type === 'checkbox' ? e.target.checked : e.target.value;

        this.setState({
            hasHeaders: value
        })

    }


    onSubmit(e){
        e.preventDefault();

        // Check the error
        var currentError = "";
        if (!this.state.filePath)
        {
            currentError = "You have not specified a file.";
        }
        if( currentError === ""){
            this.setState({error: currentError})
            this.props.onSubmit({
                filePath: this.state.filePath,
                fileType: this.state.fileType,
                hasHeaders: this.state.hasHeaders
            })
        } else {
            this.setState({error: currentError})
        }
    }

    clearError(){
        this.setState({error: ""})
    }

    render() {
        return (
            <form className="form__importData" onSubmit={this.onSubmit}>
                {this.state.error !== '' && <ErrorModal errorMessage={this.state.error} clearError={this.clearError} />}

                {/* File Type */}
                <div className="form__input">
                    <label className = "form__label">File Type</label>
                    <select value={this.state.fileType}
                            onChange={this.onFileTypeChange}
                    >
                        <option value="csv">CSV</option>
                        <option value="tsv">TSV</option>
                        <option value="bar">| separated</option>
                    </select>
                    
                </div>

                
                <div className="form__input">
                    <label className = "form__label">Check if this file has the headers</label>
                    <input name="hasHeaders" type="checkbox" checked={this.state.hasHeaders} onChange={this.handleCheckChange} />
                </div>

                {/* File Path */}
                <div className = "form__input">
                        <input id="fileInput"
                            type="file"
                            ref={(ref) => this.upload = ref}
                            onChange={this.onChangeFile}
                        />
                </div>

                <div className = "form__input center">
                    <button className="button">Import Data</button>
                </div>
            </form>
        )
    }
}