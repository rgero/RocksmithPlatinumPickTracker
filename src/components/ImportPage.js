import React from 'react';
import {connect} from 'react-redux';
import ImportForm from './ImportForm';
import ErrorModal from './ErrorModal';
import {startAddSong} from '../actions/songs';

export class ImportPage extends React.Component {

    constructor(props){
        super(props);
        this.onSubmit = this.onSubmit.bind(this);
        this.readFile = this.readFile.bind(this);
        this.processSongImport = this.processSongImport.bind(this);

        this.state = {
            fileType: "",
            hasHeaders: false,
        }
    }

    readFile = (filePath) => {
        const tempFileReader = new FileReader();
        return new Promise((resolve, reject) => {
            tempFileReader.onerror = () => {
                tempFileReader.abort();
                reject(new DOMException("Problem with file"));
            }
        
            tempFileReader.onload = () => {
                resolve(tempFileReader.result);
            }

            tempFileReader.readAsText(filePath);
        });
    }

    processSongImport(content){
        return new Promise( (resolve, reject) => {
            // Split string into array
            var contentList = content.split('\n');
            
            // Check to see up
            var songs = [];
            var index = 0;
            if (this.state.hasHeaders){
                var index = 1;
            }

            for(var i = index; i < content.length - 1; i++)
            {
                var entry = contentList[i]

                if (entry){

                    var sepCharacter = ',';
                    switch(this.state.fileType){
                        case "csv":
                            sepCharacter = ","
                            break;
                        case "tsv":
                            sepCharacter = "\t";
                            break;
                        case "bar":
                            sepCharacter = "|"
                            break;
                        default:
                            break;
                    }
                    entry = entry.split(sepCharacter);
                    var testObject = {};
                    testObject["artist"] = entry[0];
                    testObject["songName"] = entry[1];
                    testObject["date"] = entry[2];
                    testObject["path"] = entry[3];
                    testObject["difficulty"] = entry[4];
                    testObject["accuracy"] = entry[5];
                    testObject["notes"] = entry[6];
                    testObject["level"] = "Platinum";

                    songs.push(testObject);
                }

            }
            if (songs.length > 0){
                this.setState({processing: true});
                for(var currentIndex = 0; currentIndex < songs.length; currentIndex++){
                    this.setState({
                        processMessage: `Processing: ${currentIndex} of ${songs.length}`
                    })
                    this.props.addSong(songs[currentIndex]);
                }
                this.setState({
                    processing: false,
                    processMessage: ""
                })
            }

            resolve("Success");
        })
    }

    async onSubmit(fileProps){
        this.setState({
            fileType: fileProps.fileType,
            hasHeaders: fileProps.hasHeaders
        })
        try{
            const fileContents = await this.readFile(fileProps.filePath);
            const successStatus = await this.processSongImport(fileContents);
            this.props.history.push('/');
        } catch (e) {
            console.warn(e.message);
        }

    }

    render(){
        return(
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Import Data</h1>
                    </div>
                </div>
                <div className="content-container">
                    <div className="content-infobox">
                        The format should be the following:
                        <div className="content-noticebox">Artist | Song Name | Date | Path | Difficulty | Accuracy | Notes</div>
                        For this utility, the pick level "Platinum" is presumed.
                    </div>
                    <ImportForm onSubmit={this.onSubmit}/>
                </div>
            </div>
        )
    }
}


const mapDispatchToProps = (dispatch) => {
    return {
        addSong: (song) => dispatch(startAddSong(song))
    }
}

export default connect(undefined, mapDispatchToProps)(ImportPage);