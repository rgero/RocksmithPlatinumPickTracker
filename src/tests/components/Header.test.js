// react-test-renderer
import React from 'react'
import {shallow} from 'enzyme'
import {Header} from '../../components/Header';

test('Header Test', ()=>{
    const wrapper = shallow(<Header startLogout={()=>{}}/>)
    expect(wrapper).toMatchSnapshot();
})

test('Testing Log out', ()=> {
    const onClickSpy = jest.fn();
    const wrapper = shallow(<Header startLogout={onClickSpy}/>)
    const logoutButton = wrapper.find('button');
    expect(logoutButton.length).toBe(1);
    logoutButton.simulate('click');
    expect(onClickSpy).toHaveBeenCalled();
})