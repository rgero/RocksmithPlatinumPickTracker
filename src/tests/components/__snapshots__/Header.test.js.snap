// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Test 1`] = `
<header
  className="header"
>
  <div
    className="header__content"
  >
    <Link
      className="header__title"
      exact="true"
      to="/"
    >
      <h1>
        Rocksmith Pick Tracker
      </h1>
    </Link>
    <button
      className="button button--link"
      onClick={[Function]}
    >
      Log out
    </button>
  </div>
</header>
`;
